hudi-utilities/src/main/java/org/apache/hudi/utilities/keygen/MultiFormatTimestampBasedKeyGenerator.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities.keygen;
+
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hudi.DataSourceUtils;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.util.TypedProperties;
+import org.apache.hudi.exception.HoodieKeyException;
+import org.apache.hudi.exception.HoodieNotSupportedException;
+import org.apache.hudi.keygen.SimpleKeyGenerator;
+import org.apache.hudi.utilities.exception.HoodieDeltaStreamerException;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.TimeZone;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.DateTimeFormatterBuilder;
+import org.joda.time.format.DateTimeParser;
+
+/**
+ * Key generator, that relies on timestamps for partitioning field. Still picks record key by name.
+ */
+public class MultiFormatTimestampBasedKeyGenerator extends SimpleKeyGenerator {
+
+  enum TimestampType implements Serializable {
+    UNIX_TIMESTAMP, DATE_STRING, MIXED, EPOCHMILLISECONDS
+  }
+
+  private final TimestampType timestampType;
+  private final String outputDateFormat;
+  private final String configInputDateFormatList;
+  private final String configInputDateFormatDelimiter;
+
+
+  // TimeZone detailed settings reference
+  // https://docs.oracle.com/javase/8/docs/api/java/util/TimeZone.html
+  private final DateTimeZone inputDateTimeZone;
+  private final DateTimeZone outputDateTimeZone;
+
+  /**
+   * Supported configs.
+   */
+  static class Config {
+    // One value from TimestampType above
+    private static final String TIMESTAMP_TYPE_FIELD_PROP                             = "hoodie.deltastreamer.keygen.timebased.timestamp.type";
+
+    private static final String TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP                 = "hoodie.deltastreamer.keygen.timebased.input.dateformatlist";
+    private static final String TIMESTAMP_INPUT_DATE_FORMAT_LIST_DELIMITER_REGEX_PROP = "hoodie.deltastreamer.keygen.timebased.input.dateformatlistdelimiterregex";
```

Comment:
Do we really need this? The most common delimiter is going to be comma only and I believe mostly users will not be using this property.

Reply:
So, the primary reason I used this is if you look into time formats that are outside the US, many countries will use a comma in their times.  So while I don't foresee myself using it, I'm sure others would.  So while most people will take the default, it's nice to have the option and not hamstring anyone IMO.

Reply:
Looks good.